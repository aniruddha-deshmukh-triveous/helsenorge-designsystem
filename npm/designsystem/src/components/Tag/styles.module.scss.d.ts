export type Styles = {
  tag: string;
  tag__icon: string;
  'tag--banana': string;
  'tag--blueberry': string;
  'tag--cherry': string;
  'tag--emphasised': string;
  'tag--has-action': string;
  'tag--has-icon': string;
  'tag--kiwi': string;
  'tag--large': string;
  'tag--medium': string;
  'tag--neutral': string;
  'tag--normal': string;
  'tag--oncolor': string;
  'tag--plum': string;
  'tag--undo': string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
