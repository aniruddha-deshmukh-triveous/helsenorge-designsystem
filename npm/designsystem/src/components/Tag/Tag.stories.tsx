import React from 'react';
import { boolean, select, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import { withA11y } from '@storybook/addon-a11y';
import Tag, { TagAction, TagSize, TagVariant } from './Tag';
import LawBook from '../Icons/LawBook';

const stories = storiesOf('Tag', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('Default', () => (
  <div style={{ width: '40rem' }}>
    <Tag
      size={select('Size', TagSize, TagSize.medium)}
      color={select('Color', ['blueberry', 'neutral', 'cherry', 'banana', 'kiwi', 'plum'], 'blueberry')}
      variant={select('Variant', TagVariant, TagVariant.normal)}
    >
      {text('Label', 'Tekst')}
    </Tag>
  </div>
));

stories.add('With icon', () => (
  <div style={{ width: '40rem' }}>
    <Tag
      size={select('Size', TagSize, TagSize.medium)}
      color={select('Color', ['blueberry', 'neutral', 'cherry', 'banana', 'kiwi', 'plum'], 'blueberry')}
      variant={select('Variant', TagVariant, TagVariant.normal)}
      svgIcon={LawBook}
    >
      {text('Label', 'Tekst')}
    </Tag>
  </div>
));

stories.add('Action', () => (
  <div style={{ width: '40rem' }}>
    <Tag
      size={select('Size', TagSize, TagSize.medium)}
      color={select('Color', ['blueberry', 'neutral', 'cherry', 'banana', 'kiwi', 'plum'], 'blueberry')}
      variant={select('Variant', TagVariant, TagVariant.normal)}
      action={select('Action', TagAction, TagAction.remove)}
      onClick={() => console.log('Tag clicked')}
    >
      {text('Label', 'Tekst')}
    </Tag>
  </div>
));
