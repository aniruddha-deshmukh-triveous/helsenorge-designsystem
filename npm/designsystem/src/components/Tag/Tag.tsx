import React from 'react';

import cn from 'classnames';

import { AnalyticsId } from '../../constants';
import { useHover } from '../../hooks/useHover';
import { palette, PaletteNames } from '../../theme/palette';
import Icon, { IconSize, SvgIcon } from '../Icons';
import Undo from '../Icons/Undo';
import X from '../Icons/X';

import styles from './styles.module.scss';

export enum TagSize {
  medium = 'medium',
  large = 'large',
}

export enum TagAction {
  remove = 'remove',
  undo = 'undo',
}

export enum TagVariant {
  normal = 'normal',
  oncolor = 'oncolor',
  emphasised = 'emphasised',
}

export type TagColors = Extract<PaletteNames, 'blueberry' | 'neutral' | 'cherry' | 'banana' | 'kiwi' | 'plum'>;

interface TagProps {
  /** Sets the text of the tag */
  children: string;
  /** Sets the size of the tag. Default: small */
  size?: keyof typeof TagSize;
  /** Sets the background of the tag. Not used if action is "undo". Default: blueberry */
  color?: TagColors;
  /** Adds an icon to the tag. Not shown if action is set. */
  svgIcon?: SvgIcon;
  /* Changes the appearance of the tag. Not used if action is "undo". Default: normal */
  variant?: keyof typeof TagVariant;
  /* Makes the tag a clickable button that performs an action. onClick must also be set. */
  action?: keyof typeof TagAction;
  /* Called when action is set and the tag is clicked on. action must also be set. */
  onClick?: () => void;
  /** Sets the data-testid attribute on the expander button. */
  testId?: string;
}

type ActionTagProps = Required<Pick<TagProps, 'children' | 'size' | 'color' | 'variant' | 'action' | 'onClick'>> & Pick<TagProps, 'testId'>;

const ActionTag: React.FC<ActionTagProps> = props => {
  const { children, size, color, variant, action, onClick, testId } = props;

  const { hoverRef, isHovered } = useHover<HTMLButtonElement>();

  const tagClasses = cn(
    styles.tag,
    styles[`tag--${size}`],
    styles[`tag--${action}`],
    styles['tag--has-action'],
    action === 'remove' && [styles[`tag--${color}`], styles[`tag--${variant}`]]
  );

  const getActionIcon = (): SvgIcon => {
    switch (action) {
      case 'remove':
        return X;
      case 'undo':
        return Undo;
    }
  };

  return (
    <button className={tagClasses} onClick={onClick} ref={hoverRef} type="button" data-testid={testId} data-analyticsid={AnalyticsId.Tag}>
      {children}
      <Icon
        svgIcon={getActionIcon()}
        size={IconSize.XXSmall}
        color={palette[`${action === 'undo' ? 'blueberry' : color}800`]}
        isHovered={isHovered}
      />
    </button>
  );
};

const Tag: React.FC<TagProps> = props => {
  const { children, size = TagSize.medium, color = 'blueberry', svgIcon, variant = 'normal', action, onClick, testId } = props;

  const tagClasses = cn(styles.tag, styles[`tag--${size}`], styles[`tag--${color}`], styles[`tag--${variant}`], {
    [styles['tag--has-icon']]: svgIcon,
  });

  if (action && onClick) {
    return (
      <ActionTag size={size} color={color} variant={variant} action={action} onClick={onClick} testId={testId}>
        {children}
      </ActionTag>
    );
  }

  return (
    <span className={tagClasses} data-testid={testId} data-analyticsid={AnalyticsId.Tag}>
      {svgIcon && <Icon svgIcon={svgIcon} size={IconSize.XXSmall} color={palette[`${color}800`]} className={styles.tag__icon} />}
      {children}
    </span>
  );
};

export default Tag;
