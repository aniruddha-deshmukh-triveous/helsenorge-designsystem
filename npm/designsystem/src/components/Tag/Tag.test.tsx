import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Tag from './Tag';

describe('Gitt at Tag skal vises vanlig', (): void => {
  describe('Når Tagen vises', (): void => {
    test('Så er teksten synlig', (): void => {
      render(<Tag>Tekst</Tag>);

      const tag = screen.getByText('Tekst');

      expect(tag).toBeVisible();
    });
    test('Så ser den ut slik den skal', (): void => {
      const { container } = render(<Tag>Tekst</Tag>);

      expect(container).toMatchSnapshot();
    });
  });
});

describe('Gitt at Tag skal vises som remove-knapp', (): void => {
  describe('Når man klikker på knappen', (): void => {
    test('Så kalles click-funksjonen', (): void => {
      const mockClickHandler = jest.fn();

      render(
        <Tag action="remove" onClick={mockClickHandler}>
          Knapp
        </Tag>
      );

      const tag = screen.getByRole('button', { name: 'Knapp' });

      expect(tag).toBeVisible();

      userEvent.click(tag);

      expect(mockClickHandler).toHaveBeenCalledTimes(1);
    });
  });
});

describe('Gitt at Tag skal vises som undo-knapp', (): void => {
  describe('Når man klikker på knappen', (): void => {
    test('Så kalles click-funksjonen', (): void => {
      const mockClickHandler = jest.fn();

      render(
        <Tag action="undo" onClick={mockClickHandler}>
          Knapp
        </Tag>
      );

      const tag = screen.getByRole('button', { name: 'Knapp' });

      expect(tag).toBeVisible();

      userEvent.click(tag);

      expect(mockClickHandler).toHaveBeenCalledTimes(1);
    });
  });
  describe('Når onClick ikke er definert', (): void => {
    test('Så vises ikke tagen som knapp likevel', (): void => {
      render(<Tag action="undo">Knapp</Tag>);

      const tag = screen.queryByRole('button');

      expect(tag).not.toBeInTheDocument();
    });
  });
  describe('Når action ikke er definert', (): void => {
    test('Så vises ikke tagen som knapp likevel', (): void => {
      render(<Tag onClick={jest.fn()}>Knapp</Tag>);

      const tag = screen.queryByRole('button');

      expect(tag).not.toBeInTheDocument();
    });
  });
});
