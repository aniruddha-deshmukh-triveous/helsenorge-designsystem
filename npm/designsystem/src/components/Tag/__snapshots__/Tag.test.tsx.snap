// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at Tag skal vises vanlig Når Tagen vises Så ser den ut slik den skal 1`] = `
<div>
  <span
    class="tag tag--medium tag--blueberry tag--normal"
    data-analyticsid="tag"
  >
    Tekst
  </span>
</div>
`;
